### ISSUE TITLE
Expression constraint validation and invalid conceptIds

### ISSUE THREAD

**Issue by dmarkwell** [IHTSDO/snowstorm#145]
While working on the TSG documentation, I noted that the Snowstorm **concepts** service does not validate ECL expressions presented in the **ecl** parameter.
There are two aspect to this:
1. Erroneous concept ids (e.g. incorrect format or check digit) or non-existent concepts (potentially valid Id for concept not in the edition)
    * These do not result in error messages
    * The result of the call is likely to be misleading in one of the following ways:
        * an empty set or (if the erroneous conceptId is part of a required condition)
        * a set that contains only concepts arising from other clauses joined by OR to the clause with the erroneous id
        * a set that contains concepts that should have been excluded by a MINUS clause containing the erroneous id
    * The first of these erroneous results is will probably be noticed by the person constructing the constraint. However, the other two types of error are more likely to be overlooked as they return a result set which may seem superficially valid.
1. ECL syntax errors
    * These result in an error message. However no information is provided about the nature of the error
    * It might be useful to add an ECL validation service to facilitate checking of invalid ecl (and/or to include this feature in the response to a call that contains erroneous ECL.

**Comment by kaicode** [comment]
This is a tricky area. 

1. Concepts could exist in an edition or extension which have an invalid SCTID. I feel that the ECL service should not prevent a user selecting these concepts. I feel that attempting to select concepts which do not exist should result in empty results however I agree that this could get misleading if those concepts are just part of the query definition and not expected to be returned. A perfectly valid ECL query could break silently if one or more of the concepts in the query become inactive.

2. We may only be able to provide limited additional information when an ECL query can not be parsed. The ANTLR parser we use does not provide much information when things go wrong. 

We have also found that the intention of some complex ECL queries can be ambiguous if not enough parenthesis are provided. Unfortunately the current ECL language specification does not enforce where parenthesis need to be used to prevent this.

I like the idea of providing an ECL validation function or returning warnings in the response if any concepts in the request do not exist as active. An optional ECL "strict mode" could make request fail if there are any warnings.

We do not have capacity to work on this right now because we are heads down on the Elasticsearch 7 upgrade and then concrete domain support. I'll leave this issue open in case anyone in the community would like to help.

**Comment by dmarkwell** [comment]
Thanks for the response. I understand the issue related to syntax and also the priorities and time pressures etc. 
I would also support a less flexible approach to use (or non use) of brackets in ECL because that would create clearer more explicitly intelligible constraints.

However, I don't think these points undermine the case for addressing this issue in due course.
I am not convinced by the arguments about invalid SCTIDs
1. The rules on SCTID validity and check-digits have been around for 20 years now and anyone using invalid SCTIDs should not expect a SNOMED terminology service to ignore these errors
1. Typographical errors are a far more likely cause of invalid identifiers (and may also explain some conceptIds that refer to missing concepts)
1. Similarly in the case of missing concepts in a given addition, the fact that a result was computed in an Edition that does not contain a concept that is used as one of the inclusion or exclusion conditions is relevant. It may or may not negate the value of an ECL query. However, it is important information.
   * In practice a simple test of each of the conceptIds in an ECL expression to see if it exists in the terminology is extremely easy to do:
e.g. 
Call: [branch]/concepts?conceptIds=[id1],[id2],...,[idN]
Result either does or does not contain all the concepts in the conceptIds list.

So I hope at some point when other priorities have been cleared it will be possible to address this point.
 
Thanks
David

**Comment by jbarcas** [comment]
So, I see several topics here:
1. ECL validation
    * Digit checking to detect malformed conceptIds.
    * Prevent from conceptIds that, although well formed, do not belong to one given edition.
2. ECL syntax
    * Improve response by adding information about the nature of the potential errors.
    * Ambiguity in complex queries related to the use of parentheses.

Regarding the first point: IMHO the user should be informed if a SCTID doesn't belong to the edition/extension he is attemping to query against (or if a malformed conceptId is provided as input), that way the user has more feedback about its query. But if, as @kaicode said, "_Concepts could exist in an edition or extension which have an invalid SCTID_", the only validation we can perform here is to check if they are active concepts in the current edition. I'm not sure if I'm understanding the idea of returning warnings in the response but, in principle, I'm not convinced with that (nor the idea of `strictMode` flag). Could you please expand your suggestion?

Regarding the second point: due to ANTLR parser doesn't provide detailed exception messages and ECL spec is not that robust on parentheses location, it seems that it's not possible to move forward on this for now, maybe a new issue for this?

Regarding @dmarkwell comment:

> In practice a simple test of each of the conceptIds in an ECL expression to see if it exists in the terminology is extremely easy  to do:
> e.g.
> Call: [branch]/concepts?conceptIds=[id1],[id2],...,[idN]
> Result either does or does not contain all the concepts in the conceptIds list.

@kaicode, please correct me if I am wrong but, the heart of the matter here is to parse de ECL expression to retrieve the involved conceptIds.

**Comment by kaicode** [comment]
Hi @jbarcas. Yes it would be a good to parse all the concept ids from the ECL and return an error if any of them do not exist or are not active on the current branch. 

HTTP response codes 400 (bad request) and 404 (not found) seem appropriate. I am leaning towards 400 because it's the content of the request (params) rather than the resource in the URL which has the problem.

The response body should contain the set of concepts which are inactive or do not exist. Something like this would be ideal:
```
{
"error": "BAD_REQUEST",
"message": "Concepts in the ECL request do not exist or are inactive on this branch: 12341234, 34563456."
}
```

I agree that ECL syntax parsing improvement could be split into a separate ticket.
Thanks for clarifying.

**Comment by jbarcas** [comment]
Alright! completely agree with the 400 response.

You can assign me the issue if you want.
Regards

**Comment by jbarcas** [comment]
hi @kaicode, PR is open to review.

Cheers.
Jesús.

**Comment by jbarcas** [comment]
> I agree that ECL syntax parsing improvement could be split into a separate ticket.

What about this? Is it worth opening a new ticket for when you are ready to resolve it?

Cheers

**Comment by kaicode** [comment]
Fixed in [5.0.6](https://github.com/IHTSDO/snowstorm/releases/tag/5.0.6).
